src/platform/packages/private/kbn-scout-reporting/src/cli/upload_events.ts

```diff
@@ -8,60 +8,136 @@
  */
 
 import fs from 'node:fs';
+import path from 'node:path';
+
 import { Command } from '@kbn/dev-cli-runner';
 import { createFlagError } from '@kbn/dev-cli-errors';
+import { ToolingLog } from '@kbn/tooling-log';
+
 import {
   SCOUT_REPORTER_ES_URL,
   SCOUT_REPORTER_ES_API_KEY,
   SCOUT_REPORTER_ES_VERIFY_CERTS,
+  SCOUT_REPORT_OUTPUT_ROOT,
 } from '@kbn/scout-info';
 import { ScoutReportDataStream } from '../reporting/report/events';
 import { getValidatedESClient } from '../helpers/elasticsearch';
 
+const readFilesRecursively = (directory: string, callback: Function) => {
+  const files = fs.readdirSync(directory);
+  files.forEach((file) => {
+    const filePath = path.join(directory, file);
+    const stat = fs.statSync(filePath);
+    if (stat.isDirectory()) {
+      readFilesRecursively(filePath, callback);
+    } else if (stat.isFile()) {
+      callback(filePath);
+    }
+  });
+};
+
+export const uploadAllEventsFromPath = async (
+  eventLogPath: string,
+  options: {
+    esURL: string;
+    esAPIKey: string;
+    verifyTLSCerts: boolean;
+    log: ToolingLog;
+  }
+) => {
+  // Validate CLI options
+  if (!fs.existsSync(eventLogPath)) {
+    throw createFlagError(`The provided event log path '${eventLogPath}' does not exist.`);
+  }
+
+  // If the provided event log path is a file, ensure it ends with .ndjson
+  if (!fs.statSync(eventLogPath).isDirectory()) {
+    if (!eventLogPath.endsWith('.ndjson')) {
+      throw createFlagError(`The provided event log file '${eventLogPath}' must end with .ndjson.`);
+    }
+  }
+
+  // ES connection
+  options.log.info(`Connecting to Elasticsearch at ${options.esURL}`);
+  const es = await getValidatedESClient(
+    {
+      node: options.esURL,
+      auth: { apiKey: options.esAPIKey },
+      tls: {
+        rejectUnauthorized: options.verifyTLSCerts,
+      },
+    },
+    { log: options.log, cli: true }
+  );
+
+  // Event log upload
+  const reportDataStream = new ScoutReportDataStream(es, options.log);
+  let numberOfNdjsonFilesFound = 0;
+
+  if (fs.statSync(eventLogPath).isDirectory()) {
+    readFilesRecursively(eventLogPath, (filePath: string) => {
+      if (filePath.endsWith('.ndjson')) {
+        reportDataStream.addEventsFromFile(filePath);
```

Comment:
I was wondering about the order of the log output:
<img width="1777" height="569" alt="image" src="https://github.com/user-attachments/assets/79438421-bb95-41ca-83ae-710f4fa42861" />

And checking the code, it turns out that `addEventsFromFile` is actually an async method. Did you intentionally not `await` it when calling from this file (here and a few lines down for the single file)?

Reply:
Looks good now :+1: